### PR TITLE
 some line break optimization 

### PR DESCRIPTION
**Pull request by copyliu**

### CODE REVIEW

File: Engine/Entities/Font.hpp [hunk 1]
```diff
@@ -33,18 +33,23 @@ const uint32_t EmDash{u'—'};
...
-static const uint32_t NotLineEnd[]{u'\x0028', u'\x005B', u'\x007B', u'\x00B7', u'\x2018', u'\x201C', u'\x300A', u'\x300C', u'\x300E',
-                                   u'\xFF04', u'\xFF08', u'\xFF0E', u'\xFF5B'};
+inline bool IsEnLetter(uint32_t ch) {
+	return (ch >= u'a' && ch <= 'z') || (ch >= u'A' && ch <= u'Z');
+}
+inline bool IsNumber(uint32_t ch) {
+	return (ch >= '0' && ch <= u'9');
+}
+inline bool IsNumberOrLetter(uint32_t ch) {
+	return IsNumber(ch) || IsEnLetter(ch);
```
Comment (vit9696) [inline Engine/Entities/Font.hpp]: I believe this needs isRuLetter too. Also, could you please fix the case of the names? We name the functions with a small first letter.

**Comment by vit9696** [comment]
Hi, sorry for a slight delay. The changes look good to me, but I spotted a nuance making them English and Chinese specific excluding Russian. That wants to be fixed. I will also ping @knox-umineko for him to handle the review too.

**Comment by copyliu** [comment]
Russian like English can only wrap in spaces IIRC, only CJK characters can wrap in EVERY character (except for certain punctuation), but I don't know if there are special rules in Russian. Added @knox-umineko to reviewer list

**Comment by vit9696** [comment]
@copyliu I mean, you added linebreaking for En/CJK mixing for all languages. I am not sure whether it works with Russian and English, because you will make a space before Kanji anyway.

- If it does, it should work the same for both Russian and English.
- If it does not, it should be inside Japanese language if, and perhaps isNumberOrLetter should be renamed to isNumberOrEnLetter.

**Comment by copyliu** [comment]
hmm, I think it's ok linebreaking in CJK/non-CJK mixing for all languages not just English. Thanks find it.

**Comment by vit9696** [comment]
Actually the way it is now it definitely not ok, as it will line-break with punctuation symbols. To be honest, more I think, it feels like an Asian language specific feature.

**Comment by copyliu** [comment]
ahh yes. I think we can remove about CJK/non-CJK mixing support in code just do it in script, add a space between CJK/non-CJK characters, it's easier than do it in code. and it's very rare

File: Engine/Entities/Font.hpp [hunk 2]
```diff
@@ -33,18 +33,23 @@ const uint32_t EmDash{u'—'};
...
-//([{·‘“《「『＄（．｛
-static const uint32_t NotLineEnd[]{u'\x0028', u'\x005B', u'\x007B', u'\x00B7', u'\x2018', u'\x201C', u'\x300A', u'\x300C', u'\x300E',
-                                   u'\xFF04', u'\xFF08', u'\xFF0E', u'\xFF5B'};
+inline bool isEnLetter(uint32_t ch) {
+	return (ch >= u'a' && ch <= 'z') || (ch >= u'A' && ch <= u'Z');
+}
+inline bool isNumber(uint32_t ch) {
+	return (ch >= '0' && ch <= u'9');
+}
+inline bool isNumberOrLetter(uint32_t ch) {
```
Comment (vit9696) [inline Engine/Entities/Font.hpp]: Please rename to isNumberOrEnLetter. The rest looks fine to me, I will merge it around the weekend if no issues are found.

**Pull request merged**